odoo/cli/code_upgrade.py

```diff
@@ -0,0 +1,131 @@
+
+"""
+
+AVAILABLE_EXT = ('py', 'js', 'css', 'scss', 'xml', 'csv')
+
+
+class FileDescriptor:
+    addon: Path
+    path: Path
+    content: str
+
+    def __init__(self, path: Path, addon: Path) -> None:
+        self.path = path
+        self.addon = addon
+        self._content = None
+        self.dirty = False
+
+    @property
+    def content(self):
+        if self._content is None:
+            self._content = self.path.open("r").read()
+        return self._content
+
+    @content.setter
+    def content(self, value):
+        if self._content != value:
+            self._content = value
+            self.dirty = True
+
+
+class FileManager:
+    addons_path: list[str]
+    glob: str | None
+
+    def __init__(self, addons_path: list[str], glob: str = '**/*') -> None:
+        self.addons_path = addons_path
+        self.glob = glob
+        self._files = {}
+        for addon_path in addons_path:
+            for path_addon in Path(addon_path).iterdir():
+                for path in path_addon.glob(glob):
+                    if path.is_file() and path.name.rsplit('.')[-1] in AVAILABLE_EXT:
+                        self._files[str(path)] = FileDescriptor(path, path_addon)
+
+    def __iter__(self) -> Iterator[FileDescriptor]:
+        return iter(self._files.values())
+
+    def get_file(self, path):
+        return self._files.get(str(path))
+
+
+def get_version(value: str) -> tuple[int, ...]:
+    return tuple(int(x) if x.isnumeric() else x for x in value.split('.'))
+
+
+def get_upgrade_code_scripts(from_version: tuple[int, ...], to_version: tuple[int, ...]) -> list[ModuleType]:
+    modules = []
+    script_paths = list(Path(__file__).parent.parent.glob('code_upgrade/*.py'))
+    script_paths.sort(key=str)
+    for script_path in script_paths:
+        version = get_version(script_path.name.split('-', 1)[0])
+        if from_version <= version <= to_version:
+            module = SourceFileLoader(script_path.name, str(script_path)).load_module()
+            modules.append((script_path.name, module))
+    return modules
+
+
+def migrate(
+        from_version: tuple[int, ...],
+        to_version: tuple[int, ...],
+        addons_path: list[str],
+        glob: str | None,
+        test: bool = False):
+
+    modules = get_upgrade_code_scripts(from_version, to_version)
+    file_manager = FileManager(addons_path, glob)
+    for (name, module) in modules:
+        print(f'update script: {name}')  # noqa: T201
+        module.upgrade(file_manager)
+
+    for file in file_manager:
+        if file.dirty:
+            print('updated: ', file.path)  # noqa: T201
+            if not test:
+                file.path.open("w").write(file.content)
+
+
+class Code_Upgrade(Command):
+    def run(self, cmdargs):
+        odoo = Path(__file__).parent.parent.parent
+        to_version = SourceFileLoader('release', str(next(odoo.glob('odoo/release.py')))).load_module().version_info
```

Comment:
You can just import `odoo.release`.

Reply:
we don't want to import odoo into this file